File: ruby/grains/grains.rb
class Grains
  FIRST_SQUARE = 1
  LAST_SQUARE  = 64

  def self.square(n)
    if n < FIRST_SQUARE || n > LAST_SQUARE
      raise ArgumentError, "#{n} is invalid. Must be between " \
                           "#{FIRST_SQUARE} and #{LAST_SQUARE}"
    end
    number_of_grains(n)
  end

  def self.total
    FIRST_SQUARE.upto(LAST_SQUARE).reduce(0) do |total_grains, n|
      total_grains += square(n)
    end
  end

  def self.number_of_grains(n)
    2**(n - 1)
  end

  private_class_method :number_of_grains
end

File: javascript/grains/grains.js
const BigI = require('./big-integer');

class Grains {
  square(n) {
    return this.num_grains_in_square(n).toString();
  }

  num_grains_in_square(n) {
    return BigI(2).pow(n - 1);
  }

  total() {
    let total = BigI();
    for (let n = 1; n <= 64; n++) {
      total = total.add(this.num_grains_in_square(n));
    }
    return total.toString();
  }
}

module.exports = Grains;
